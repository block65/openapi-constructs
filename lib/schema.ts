import { Construct } from 'constructs';
import type { oas31 } from 'openapi3-ts';

interface SchemaOptions<T extends oas31.SchemaObject> {
  schema: T;
}

export class Schema<
  T extends oas31.SchemaObject = oas31.SchemaObject,
> extends Construct {
  private options: SchemaOptions<T>;

  constructor(scope: Construct, id: string, options: SchemaOptions<T>) {
    super(scope, id);

    this.options = options;
  }

  public get schema() {
    return this.options.schema;
  }

  public get schemaKey() {
    return this.node.id;
  }

  public jsonPointer() {
    return `#/components/schemas/${this.schemaKey}`;
  }

  public referenceObject(): oas31.ReferenceObject {
    return {
      $ref: this.jsonPointer(),
    };
  }

  // eslint-disable-next-line class-methods-use-this
  public validate() {
    return [];
  }

  public synth() {
    return this.options.schema;
  }
}
