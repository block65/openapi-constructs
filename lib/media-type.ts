import { Construct } from 'constructs';
import type { oas31 } from 'openapi3-ts';
import { Reference } from './reference.js';
import type { Schema } from './schema.js';

export interface MediaTypeOptions {
  contentType:
    | 'multipart/form-data'
    | 'application/x-www-form-urlencoded'
    | 'application/json'
    | 'image/*';
  schema: Schema | Reference<Schema>;
}

export class MediaType extends Construct {
  private options: MediaTypeOptions;

  public get contentType() {
    return this.options.contentType;
  }

  constructor(scope: Construct, id: string, options: MediaTypeOptions) {
    super(scope, id);
    this.options = options;
  }

  public synth(): oas31.MediaTypeObject {
    return {
      schema:
        this.options.schema instanceof Reference
          ? this.options.schema.synth()
          : this.options.schema.referenceObject(),
    };
  }
}
