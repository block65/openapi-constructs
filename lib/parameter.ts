import { Construct } from 'constructs';
import type { OpenAPIV3 } from 'openapi-types';
import type { Api } from './api.js';
import type { Schema } from './schema.js';

interface ParameterOptionsBase<TName extends string | number | symbol> {
  name: TName;
  in: 'query' | 'header' | 'path' | 'cookie';
  required: boolean;
  description?: string;
  deprecated?: boolean;
  allowEmptyValue?: boolean;
  style?: 'simple';
}

interface ParameterOptions<TName extends string | number | symbol>
  extends ParameterOptionsBase<TName> {
  schema: Schema;
}

// interface ParameterOptions<TName extends string | number | symbol>
//   extends ParameterOptionsBase<TName> {
//   content: unknown;
// }

export class Parameter<
  TName extends string | number | symbol = '',
> extends Construct {
  private options: ParameterOptions<TName>;

  constructor(scope: Api, id: string, options: ParameterOptions<TName>) {
    super(scope, id);
    this.options = options;
  }

  public referenceObject(): OpenAPIV3.ReferenceObject {
    return {
      $ref: this.jsonPointer(),
    };
  }

  public get schemaKey() {
    return this.node.id;
  }

  public jsonPointer(): string {
    return `#/components/parameters/${this.schemaKey}`;
  }

  public synth(): OpenAPIV3.ParameterObject | OpenAPIV3.ReferenceObject {
    return {
      name: this.options.name.toString(),
      in: this.options.in,
      required: this.options.required,
      ...(this.options.allowEmptyValue && {
        allowEmptyValue: this.options.allowEmptyValue,
      }),
      ...(this.options.style && { style: this.options.style }),
      ...(this.options.schema && {
        schema: this.options.schema.referenceObject(),
      }),
    };
  }
}
