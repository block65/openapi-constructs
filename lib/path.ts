import { Construct } from 'constructs';
import type { oas31 } from 'openapi3-ts';
import type { Api } from './api.js';
import { HttpMethods } from './http-methods.js';
import { Operation, type OperationOptions } from './operation.js';
import type { Parameter } from './parameter.js';
import type { Server } from './server.js';
import type { Tag } from './tag.js';
import type { ExtractRouteParams } from './types.js';

interface PathOptions<TPath extends string> {
  path: TPath;
  summary?: string;
  servers?: Server[];
  parameters?: Parameter<keyof ExtractRouteParams<TPath>>[];
  tags?: Set<Tag>;
}

export class Path<TPath extends string = '/'> extends Construct {
  public options: PathOptions<TPath>;

  constructor(scope: Api, options: PathOptions<TPath>) {
    super(scope, options.path);
    this.options = options;
  }

  public addOperation(
    method: HttpMethods,
    options: OperationOptions<TPath>,
  ): this {
    // make sure we are not duplicating tags
    // options.tags?.forEach((tag) => strict(!this.options.tags?.has(tag)));

    // eslint-disable-next-line no-new
    new Operation(this, method, {
      ...options,
      tags: new Set([...(this.options.tags || []), ...(options.tags || [])]),
    });
    return this;
  }

  public get schemaKey() {
    return this.options.path;
  }

  public synth() {
    return {
      ...Object.fromEntries(
        this.node
          .findAll()
          .filter(
            (child): child is Operation<TPath> => child instanceof Operation,
          )
          .sort((a, b) => a.order - b.order)
          .map((child) => [child.method, child.synth()]),
      ),
      ...(this.options.servers && {
        servers: this.options.servers.map((child) => child.synth()),
      }),
      ...(this.options.parameters && {
        parameters: this.options.parameters.map((child) =>
          child.referenceObject(),
        ),
      }),
      ...(this.options.summary && { summary: this.options.summary }),
    } satisfies oas31.PathItemObject;
  }
}
