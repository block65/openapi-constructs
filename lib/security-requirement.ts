import { Construct } from 'constructs';
import type { OpenAPIV3_1 } from 'openapi-types';
import type { Api } from './api.js';
import type { SecurityScheme } from './security-scheme.js';

interface SecurityRequirementOptions {
  securityScheme?: SecurityScheme;
  scopes?: string[];
}

export class SecurityRequirement extends Construct {
  private options: SecurityRequirementOptions;

  constructor(
    scope: Api,
    id: string,
    options: SecurityRequirementOptions = {},
  ) {
    super(scope, id);
    this.options = options;
  }

  public synth(): OpenAPIV3_1.SecurityRequirementObject {
    return this.options.securityScheme
      ? {
          [this.options.securityScheme.node.id]: this.options.scopes || [],
        }
      : {};
  }
}
