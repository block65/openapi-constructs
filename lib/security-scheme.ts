import { Construct } from 'constructs';
import type { OpenAPIV3_1 } from 'openapi-types';

export class SecurityScheme extends Construct {
  private options: OpenAPIV3_1.SecuritySchemeObject;

  constructor(
    scope: Construct,
    id: string,
    options: OpenAPIV3_1.SecuritySchemeObject,
  ) {
    super(scope, id);
    this.options = options;
  }

  public synth(): OpenAPIV3_1.SecuritySchemeObject {
    return this.options;
  }
}
