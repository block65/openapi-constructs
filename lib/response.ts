import { Construct } from 'constructs';
import type { OpenAPIV3_1 } from 'openapi-types';
import type { Header } from './header.js';
import { MediaType, MediaTypeOptions } from './media-type.js';

interface ResponseOptions {
  content: MediaType | MediaTypeOptions;
  description?: string;
  headers?: Header[];
}

export class Response extends Construct {
  private options: ResponseOptions;

  private content: MediaType;

  constructor(scope: Construct, id: string, options: ResponseOptions) {
    super(scope, id);
    this.options = options;

    this.content =
      options.content instanceof MediaType
        ? options.content
        : new MediaType(this, `${id}MediaType`, options.content);
  }

  public synth(): OpenAPIV3_1.ResponseObject {
    return {
      description: this.options.description || '',
      content: {
        [this.options.content.contentType]: this.content.synth(),
      },
    };
  }
}
