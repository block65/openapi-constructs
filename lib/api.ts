import type { JSONSchema7 } from 'json-schema';
import type { oas31 } from 'openapi3-ts';
import { ApiLowLevel } from './ApiLowLevel.js';
import { Parameter } from './parameter.js';
import { Path } from './path.js';
import { Reference } from './reference.js';
import { Schema } from './schema.js';
import { SecurityRequirement } from './security-requirement.js';
import { SecurityScheme } from './security-scheme.js';
import { Server } from './server.js';
import { Tag } from './tag.js';

export enum OpenApiVersion {
  // V2 = '2.0',
  V3 = '3.0',
  V3_1 = '3.1.0',
}

export interface ApiOptions {
  openapi: OpenApiVersion; //  | `${OpenApiVersion}`;
  info: oas31.InfoObject;
}

export class Api extends ApiLowLevel {
  private options: ApiOptions;

  constructor(options: ApiOptions) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-explicit-any
    super(undefined as any, '');

    this.options = options;
  }

  public synth() {
    return {
      openapi: this.options.openapi,
      info: this.options.info,
      servers: this.node.children
        .filter((child): child is Server => child instanceof Server)
        .map((child) => child.synth()),
      tags: this.node.children
        .filter((child): child is Tag => child instanceof Tag)
        .map((child) => child.synth()),
      components: {
        schemas: Object.fromEntries(
          this.node.children
            .filter(
              (child): child is Schema | Reference<Schema> =>
                child instanceof Schema || child instanceof Reference,
            )
            .map((child) => [child.schemaKey, child.synth()]),
        ),
        parameters: Object.fromEntries(
          this.node.children
            .filter((child): child is Parameter => child instanceof Parameter)
            .map((child) => [child.schemaKey, child.synth()]),
        ),
        securitySchemes: Object.fromEntries(
          this.node.children
            .filter(
              (child): child is SecurityScheme =>
                child instanceof SecurityScheme,
            )
            .map((child) => [child.schemaKey, child.synth()]),
        ),
      },
      paths: Object.fromEntries(
        this.node.children
          .filter((child): child is Path => child instanceof Path)
          .map((child) => [child.schemaKey, child.synth()]),
      ),
      security: this.node.children
        .filter(
          (child): child is SecurityRequirement =>
            child instanceof SecurityRequirement,
        )
        .map((child) => child.synth()),
    } satisfies oas31.OpenAPIObject;
  }

  public synthJsonSchema() {
    return {
      $schema: 'http://json-schema.org/draft-07/schema#',
      definitions: Object.fromEntries(
        this.node.children
          .filter((child): child is Schema => child instanceof Schema)
          .map((child) => [child.schemaKey, child.synth()]),
      ),
    } satisfies JSONSchema7;
  }
}
