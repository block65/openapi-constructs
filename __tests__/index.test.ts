/* eslint-disable no-new */
import { writeFile } from 'node:fs/promises';
import SwaggerParser from '@apidevtools/swagger-parser';
import { describe, test } from '@jest/globals';
import { OpenAPIV3 } from 'openapi-types';
import yaml from 'yaml';
import { Api } from '../lib/api.js';
import { OpenApiVersion } from '../lib/index.js';
import { MediaType } from '../lib/media-type.js';
import { Parameter } from '../lib/parameter.js';
import { Path } from '../lib/path.js';
import { Response } from '../lib/response.js';
import { Schema } from '../lib/schema.js';
import { SecurityRequirement } from '../lib/security-requirement.js';
import { SecurityScheme } from '../lib/security-scheme.js';
import { Server } from '../lib/server.js';
import { Tag } from '../lib/tag.js';

describe('Basic', () => {
  test('Basic', async () => {
    const api = new Api({
      openapi: OpenApiVersion.V3_1,
      info: {
        title: 'Example REST API',
        version: '1.0.0',
      },
    });

    new Server(api, 'ExampleServer', {
      url: new URL('https://api.example.com'),
    });

    const httpBearerJwtScheme = new SecurityScheme(api, 'HttpBearerJwtScheme', {
      type: 'http',
      scheme: 'bearer',
      bearerFormat: 'JWT',
    });

    new SecurityRequirement(api, 'AllScopes', {
      securityScheme: httpBearerJwtScheme,
      scopes: [],
    });

    const userTag = new Tag(api, 'UserTag', {
      name: 'user',
    });

    const userDeleteScopeReq = new SecurityRequirement(api, 'UserReadScope', {
      securityScheme: httpBearerJwtScheme,
      scopes: ['users.delete'],
    });

    const noSecurityRequirement = new SecurityRequirement(api, 'NoSecurity');

    const addressSchema = new Schema(api, 'Address', {
      schema: {
        type: 'object',
        required: ['name'],
        additionalProperties: false,
        properties: {
          postcode: {
            type: 'integer',
            format: 'int32',
            minimum: 1000,
            maximum: 9999,
          },
        },
      },
    });

    const idSchema = new Schema(api, 'Id', {
      schema: {
        type: 'string',
      },
    });

    const user = new Schema(api, 'User', {
      schema: {
        type: 'object',
        required: ['name'],
        additionalProperties: false,
        properties: {
          name: {
            type: 'string',
          },
          address: addressSchema.referenceObject(),
          age: {
            type: 'integer',
            format: 'int32',
            minimum: 0,
          },
        },
      },
    });

    const updateUserRequest = new Schema(api, 'UpdateUserRequest', {
      schema: {
        type: 'object',
        required: ['name'],
        additionalProperties: false,
        properties: {
          name: {
            type: 'string',
          },
          address: addressSchema.referenceObject(),
          age: {
            type: 'integer',
            format: 'int32',
            minimum: 0,
          },
        },
      },
    });

    const users = new Schema(api, 'Users', {
      schema: {
        type: 'array',
        // additionalItems: false,
        uniqueItems: true,
        items: user.referenceObject(),
      },
    });

    /* const errorSchema = new Schema(api, 'ErrorSchema', {
      schema: {
        type: 'object',
        required: ['name'],
        additionalProperties: false,
        properties: {
          name: {
            type: 'string',
          },
        },
      },
    }); */

    /* const idParameter = new Parameter(api, 'IdParam', {
      name: 'userId',
      in: 'path',
      required: true,
      schema: idSchema,
    }); */

    const userIdParameter = new Parameter(api, 'UserId', {
      name: 'userId',
      in: 'path',
      required: true,
      schema: idSchema,
    });

    /* const userIdentifiersSchema = new Schema(api, 'UserIdentifiers', {
      schema: {
        type: 'object',
        additionalProperties: false,
        required: ['userId'],
        properties: {
          userId: userIdParameter.referenceObject(),
        },
      },
    }); */

    /* const errorResponse = new Response(api, 'ErrorResponse', {
      description: 'Error response',
      content: new MediaType(api, 'Error', {
        contentType: 'application/json',
        schema: errorSchema,
      }),
    }); */

    new Path(api, {
      path: '/users',
    }).addOperation(OpenAPIV3.HttpMethods.GET, {
      operationId: 'listUsers',
      tags: [userTag],
      responses: {
        200: new Response(api, 'Kek1', {
          description: 'User 200 response',
          content: new MediaType(api, 'UsersMediaType', {
            contentType: 'application/json',
            schema: users,
          }),
        }),
        // default: errorResponse,
      },
    });

    new Path(api, {
      path: '/users/{userId}',
      parameters: [userIdParameter],
      tags: new Set([userTag]),
    })
      .addOperation(OpenAPIV3.HttpMethods.GET, {
        operationId: 'getUserByIdCommand',
        responses: {
          200: new Response(api, 'GetUserById', {
            description: 'User 200 response',
            content: {
              contentType: 'application/json',
              schema: user,
            },
          }),
        },
      })
      .addOperation(OpenAPIV3.HttpMethods.DELETE, {
        operationId: 'deleteUserByIdCommand',
        security: userDeleteScopeReq,
      })
      .addOperation(OpenAPIV3.HttpMethods.HEAD, {
        operationId: 'checkUserIdAvailableCommand',
        security: noSecurityRequirement,
      })
      .addOperation(OpenAPIV3.HttpMethods.POST, {
        operationId: 'updateUserCommand',
        requestBody: {
          content: {
            contentType: 'application/json',
            schema: updateUserRequest,
          },
        },
        responses: {
          200: new Response(api, 'UpdateUserResponse200', {
            content: {
              contentType: 'application/json',
              schema: user,
            },
          }),
        },
      });

    expect(api.synth()).toMatchInlineSnapshot(`
      Object {
        "components": Object {
          "parameters": Object {
            "UserId": Object {
              "in": "path",
              "name": "userId",
              "required": true,
              "schema": Object {
                "type": "string",
              },
            },
          },
          "schemas": Object {
            "Address": Object {
              "additionalProperties": false,
              "properties": Object {
                "postcode": Object {
                  "format": "int32",
                  "maximum": 9999,
                  "minimum": 1000,
                  "type": "integer",
                },
              },
              "required": Array [
                "name",
              ],
              "type": "object",
            },
            "Id": Object {
              "type": "string",
            },
            "UpdateUserRequest": Object {
              "additionalProperties": false,
              "properties": Object {
                "address": Object {
                  "$ref": "#/components/schemas/Address",
                },
                "age": Object {
                  "format": "int32",
                  "minimum": 0,
                  "type": "integer",
                },
                "name": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "name",
              ],
              "type": "object",
            },
            "User": Object {
              "additionalProperties": false,
              "properties": Object {
                "address": Object {
                  "$ref": "#/components/schemas/Address",
                },
                "age": Object {
                  "format": "int32",
                  "minimum": 0,
                  "type": "integer",
                },
                "name": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "name",
              ],
              "type": "object",
            },
            "Users": Object {
              "items": Object {
                "$ref": "#/components/schemas/User",
              },
              "type": "array",
              "uniqueItems": true,
            },
          },
          "securitySchemes": Object {
            "HttpBearerJwtScheme": Object {
              "bearerFormat": "JWT",
              "scheme": "bearer",
              "type": "http",
            },
          },
        },
        "info": Object {
          "title": "Example REST API",
          "version": "1.0.0",
        },
        "openapi": "3.1.0",
        "paths": Object {
          "/users": Object {
            "get": Object {
              "operationId": "listUsers",
              "responses": Object {
                "200": Object {
                  "content": Object {
                    "application/json": Object {
                      "schema": Object {
                        "$ref": "#/components/schemas/Users",
                      },
                    },
                  },
                  "description": "User 200 response",
                },
              },
              "tags": Array [
                "user",
              ],
            },
          },
          "/users/{userId}": Object {
            "delete": Object {
              "operationId": "deleteUserByIdCommand",
              "security": Array [
                Object {
                  "HttpBearerJwtScheme": Array [
                    "users.delete",
                  ],
                },
              ],
              "tags": Array [
                "user",
              ],
            },
            "get": Object {
              "operationId": "getUserByIdCommand",
              "responses": Object {
                "200": Object {
                  "content": Object {
                    "application/json": Object {
                      "schema": Object {
                        "$ref": "#/components/schemas/User",
                      },
                    },
                  },
                  "description": "User 200 response",
                },
              },
              "tags": Array [
                "user",
              ],
            },
            "head": Object {
              "operationId": "checkUserIdAvailableCommand",
              "security": Array [
                Object {},
              ],
              "tags": Array [
                "user",
              ],
            },
            "parameters": Array [
              Object {
                "$ref": "#/components/parameters/UserId",
              },
            ],
            "post": Object {
              "operationId": "updateUserCommand",
              "requestBody": Object {
                "content": Object {
                  "application/json": Object {
                    "schema": Object {
                      "$ref": "#/components/schemas/UpdateUserRequest",
                    },
                  },
                },
                "description": "postRequestBody",
              },
              "responses": Object {
                "200": Object {
                  "content": Object {
                    "application/json": Object {
                      "schema": Object {
                        "$ref": "#/components/schemas/User",
                      },
                    },
                  },
                  "description": "",
                },
              },
              "tags": Array [
                "user",
              ],
            },
          },
        },
        "security": Array [
          Object {
            "HttpBearerJwtScheme": Array [],
          },
          Object {
            "HttpBearerJwtScheme": Array [
              "users.delete",
            ],
          },
          Object {},
        ],
      }
    `);

    await expect(SwaggerParser.validate(api.synth())).resolves.toBeTruthy();

    await writeFile(
      new URL('./example.json', import.meta.url),
      JSON.stringify(api.synth(), null, 2),
    );
    await writeFile(
      new URL('./example.yaml', import.meta.url),
      yaml.stringify(api.synth()),
    );
  });
});
