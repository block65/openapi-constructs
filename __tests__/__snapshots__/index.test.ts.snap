// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Synth > JSON Schema snapshot 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "Address": {
      "additionalProperties": false,
      "properties": {
        "postcode": {
          "format": "int32",
          "maximum": 9999,
          "minimum": 1000,
          "type": "integer",
        },
      },
      "required": [
        "name",
      ],
      "type": "object",
    },
    "Id": {
      "type": "string",
    },
    "UpdateUserRequest": {
      "additionalProperties": false,
      "minProperties": 1,
      "properties": {
        "address": {
          "$ref": "#/components/schemas/Address",
        },
        "age": {
          "format": "int32",
          "minimum": 0,
          "type": "integer",
        },
      },
      "type": "object",
    },
    "User": {
      "additionalProperties": false,
      "properties": {
        "address": {
          "$ref": "#/components/schemas/Address",
        },
        "age": {
          "format": "int32",
          "minimum": 0,
          "type": "integer",
        },
        "name": {
          "type": "string",
        },
        "userId": {
          "$ref": "#/components/schemas/Id",
        },
      },
      "required": [
        "name",
      ],
      "type": "object",
    },
    "Users": {
      "items": {
        "$ref": "#/components/schemas/User",
      },
      "type": "array",
      "uniqueItems": true,
    },
  },
}
`;
